### ISSUE TITLE
Rethink score aggregation logic

### ISSUE THREAD

**Issue by tju01** [FastEval/FastEval#61]
### 1. The requirement for score aggregation

Right now, there are four different benchmarks on the leaderboard. In addition, benchmarks like CoT are also composed of multiple sub-benchmarks (GSM8K + BBH + MMLU). Yet we want a single aggregate CoT score and we also want a total score across benchmarks or at least some total ranking to compare models.

So how do we aggregate the scores? Let's have a look at how some other leaderboards do it:
1. The [Open LLM Leaderboard](https://huggingface.co/spaces/HuggingFaceH4/open_llm_leaderboard) just computes a simple average without any sort of normalization before averaging. This mostly works fine because the individual numbers (ARC, HellaSwag, MMLU, TruthfulQA) are mostly in the same range.
2. The [LMSys leaderboard](https://huggingface.co/spaces/lmsys/chatbot-arena-leaderboard) doesn't compute a total score. They have three columns for scores (Arena Elo rating, MT-Bench, MMLU) and by default the Arena Elo rating is used to sort models.
3. Same for the [HumanEval+ leaderboard](https://evalplus.github.io/leaderboard.html) which focuses on "pass@1*" though here it is more acceptable since the different rankings are mostly the same.
4. Other leaderboards like [AlpacaEval](https://tatsu-lab.github.io/alpaca_eval/) only have a single score in the first place and don't need any aggregation.

The problem with our leaderboard is that it combines multiple scores that have quite different ranges. Of course it would be easy to normalize for those theoretically possible ranges before aggregating, but I think this is also not a good approach.

To explain, consider that MT-Bench can theoretically go from 1 to 10 and HumanEval+ can go from 0 to 1. However, right now the worst model on the leaderboard still achieves 3.88 on MT-Bench, yet it scores 0.00 on HumanEval+. Same for the upper range where the best model scores 8.22 on MT-Bench but only 0.67 on HumanEval+. I therefore think that using the theoretical score range (1-10 & 0-1) is not good enough and it's better to use the actually achieved scores of the worst & best model per benchmark to normalize all of the scores for every benchmark before averaging. Doing this results in total scores that I believe align more with how a human would rank the models which in the end is basically what we want.

Therefore, this is what I'm currently doing. Another very nice aspect of this is that we can get the colors on the website which makes good & bad scores _a lot_ more obvious than simply having the numbers without any colors. E.g. the good HumanEval+ score of WizardCoder is immediately very obvious with the colors, but when I didn't have those colors it wasn't that obvious without looking at all of the numbers in the table. Colors for good & bad scores really help _a lot_.

### 2. The first problem: Aggregate scores change when new models are added

The biggest problem with the normalization depending on other models is that the scores can now change if another model is added that changes the range. For example, [there is a branch with GPT-4](https://fasteval.github.io/FastEval/#?branch=gpt4) where some scores are different due to the normalization because GPT-4 changes the maximum practically achievable score for almost all benchmarks. The total ranking changes only slightly, but the total score does change quite a lot.

This is a problem because people might want to announce total scores of their model or things like that. If those scores change over time, then that's not good. This is the main problem. Now, they can still announce the individual scores like MT-Bench scores and HumanEval+ scores. But not the total aggregate score. And also not the total CoT score at least right now because that also does some normalization & aggregation inside.

### 3. The second problem: Score aggregation is implemented in JS

Another smaller problem is also that the normalization & aggregation logic is implemented in the website, i.e. in javascript right now. The reason for this is that doing it in python would mean that there would be some files that would contain the relative scores (i.e. after the normalization) and when a new model is added that changes the normalization range, all of those files would change. Which would lead to merge conflicts if there are multiple branches where people add models, e.g. if people submit their model results with a pull request.

But this is a smaller problem than the main problem with aggregate scores changing when new models are added. Though it's still worth thinking about how the general architecture should be. I would really prefer to have the score aggregation logic implemented in python rather than JS, but doing it naively would lead to merge conflicts. So it would require a different architecture.

**Comment by tju01** [comment]
The final score of a model should not depend on the other models. Therefore, what I'm going to do is look at the current minimum & maximum scores per benchmark on the current leaderboard with the current models right now and hardcode those min & max in the code to use for adjusting the scores before aggregating them.

In addition, instead of normalization using min & max I think using weights is the more common approach. It is simpler because only a single weight is needed per benchmark and it is more flexible because this weight can also take other things into account like when we deliberately want to weight a benchmark less if we think it doesn't matter that much. The only con is that total scores now won't start at close to 0 and go up close to 1, but they might be in some range like 30-50 or something like that (I don't know the exact numbers right now, we will have to see). The final scores in the range 0 to 1 are of course nicer, but I don't think it's _that_ much of an issue and other leaderboards are also like this. And stable total scores are the most important.

In addition, I'm also going to move all of the score aggregation logic from the website to python. That's really where it belongs. The website should just be a simple UI that enables looking at the scores & model outputs.

**Comment by tju01** [comment]
The colors should still depend on the minimum & maximum score though, so there needs to be some logic for that still remaining.

**Comment by tju01** [comment]
The following is documentation as to how I arrive at the various weights:

CoT > BBH:
- The total BBH score is calculated as an unweighted average of the individual scores. This follows [this paper](https://arxiv.org/abs/2210.09261)
- The minimum score with the current models on the leaderboard is 0.08627450980392158
- The maximum is 0.8509803921568627
- The current models on the leaderboard therefore cover the range 0.8509803921568627 - 0.08627450980392158 = 0.7647058823529411

CoT > MMLU:
- The total MMLU score is again an unweighted average like everyone else does it
- The minimum score is 0.09649122807017542
- The maximum is 0.7877192982456137
- Range = 0.7877192982456137 - 0.09649122807017542 = 0.6912280701754383

CoT > GSM8K:
- The minimum score is 0.016
- The maximum is 0.882
- Range = 0.882 - 0.016 = 0.866

Therefore, in order to make all of these tasks count the same and considering the range of scores that are actually covered in practice, the following weights will be used for computing the total CoT score:
- BBH: 1 / 0.7647058823529411 = 1.3076923076923077
- MMLU: 1 / 0.6912280701754383 = 1.4467005076142139
- GSM8K: 1 / 0.866 = 1.1547344110854503

These weights can now also be modified if we would consider some tasks to be more important than others. But for now I'm going to hardcode these weights.

**Comment by tju01** [comment]
The scores for MT-Bench, HumanEval+ and LM-Eval already do not depend on the other models, so the only thing left is the total score across all the four benchmarks as well as the resulting total rank.

**Comment by tju01** [comment]
Here are the minimum scores with the current models on the leaderboard:

```
{
    "mt-bench": 3.5562500000000004,
    "cot": 0.07756334159345821,
    "human-eval-plus": 0.004065040650406503,
    "lm-evaluation-harness": 48.43621488732503
}
```

And the maximum scores:

```
{
    "mt-bench": 8.89375,
    "cot": 0.8367315727010177,
    "human-eval-plus": 0.8008130081300812,
    "lm-evaluation-harness": 74.4085602041611
}
```

The resulting ranges:

```
{
    "mt-bench": 5.3375,
    "cot": 0.7591682311075595,
    "human-eval-plus": 0.7967479674796747,
    "lm-evaluation-harness": 25.972345316836076
}
```

The resulting weights (unnormalized):

```
{
    "mt-bench": 0.18735362997658078,
    "cot": 1.3172310945376207,
    "human-eval-plus": 1.2551020408163267,
    "lm-evaluation-harness": 0.03850249131532103
}
```

And normalized:

```
{
    "mt-bench": 0.06695531030705157,
    "cot": 0.4707441040341093,
    "human-eval-plus": 0.4485407975301857,
    "lm-evaluation-harness": 0.013759788128653397
}
```

**Comment by tju01** [comment]
The original problem with how scores are aggregated has been solved. The problem with the rankings being implemented in javascript still exists though, see #63
